pkg/license/license.go

```diff
@@ -0,0 +1,329 @@
+			regexp.MustCompile(`(?i)^` + copyingRegex.String() + preferredExtRegex.String() + `$`):             0.85, // COPYING.md
+			regexp.MustCompile(`(?i)^` + licenseRegex.String() + anyExtRegex.String() + `$`):                   0.80, // LICENSE.textile
+			regexp.MustCompile(`(?i)^` + copyingRegex.String() + anyExtRegex.String() + `$`):                   0.75, // COPYING.textile
+			regexp.MustCompile(`(?i)^` + licenseRegex.String() + `[-_][^.]*` + anyExtRegex.String() + `?$`):    0.70, // LICENSE-MIT
+			regexp.MustCompile(`(?i)^` + copyingRegex.String() + `[-_][^.]*` + anyExtRegex.String() + `?$`):    0.65, // COPYING-MIT
+			regexp.MustCompile(`(?i)^\w+[-_]` + licenseRegex.String() + `[^.]*` + anyExtRegex.String() + `?$`): 0.60, // MIT-LICENSE-MIT
+			regexp.MustCompile(`(?i)^\w+[-_]` + copyingRegex.String() + `[^.]*` + anyExtRegex.String() + `?$`): 0.55, // MIT-COPYING
+			regexp.MustCompile(`(?i)^` + oflRegex.String() + preferredExtRegex.String()):                       0.50, // OFL.md
+			regexp.MustCompile(`(?i)^` + oflRegex.String() + anyExtRegex.String()):                             0.45, // OFL.textile
+			regexp.MustCompile(`(?i)^` + oflRegex.String() + `$`):                                              0.40, // OFL
+			regexp.MustCompile(`(?i)^` + patentsRegex.String() + `$`):                                          0.35, // PATENTS
+			regexp.MustCompile(`(?i)^` + patentsRegex.String() + anyExtRegex.String() + `$`):                   0.30, // PATENTS.txt
+		}
+	)
+
+	var licenseFiles []LicenseFile
+	var ignore bool
+
+	err := fs.WalkDir(fsys, ".", func(filePath string, info fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() {
+			return nil
+		}
+
+		// Let's ignore all files in the melange-out/ directory, as it's not part of the source
+		if strings.Contains(filePath, "melange-out") {
+			return nil
+		}
+
+		// Check if the file matches any of the license-related regex patterns
+		for regex, weight := range filenameRegexes {
+			if regex.MatchString(info.Name()) {
+				for _, ext := range ignoredExt {
+					if ignore = strings.HasSuffix(info.Name(), ext); ignore {
+						break
+					}
+				}
+				if ignore {
+					continue
+				}
+
+				// Licenses in the top level directory have a higher weight
+				if filepath.Dir(filePath) == "." {
+					weight += 0.5
+				}
+
+				licenseFiles = append(licenseFiles, LicenseFile{
+					Name:   info.Name(),
+					Path:   filePath,
+					Weight: weight,
+				})
+				break
+			}
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Sort the license files by their 'score' (weight)
+	sort.Slice(licenseFiles, func(i, j int) bool {
+		return licenseFiles[i].Weight > licenseFiles[j].Weight
+	})
+
+	return licenseFiles, nil
+}
+
+// melangeLicenseToLicense is a helper that converts a melange license info to a License struct.
+func melangeLicenseToLicense(license string, licensePath string) License {
```

Comment:
fwiw, something like `buildLicense()` would be a more explanatory name for me. And do we need the `melange` prefix in the name?

Reply:
I named it this way since that helper is only used when we convert melange configuration's licensing information to the `License` type. I would be worried that `buildLicense()` would give the impression that this is a function that builds the License type itself, while it's meant to be a quick helper that simply fills out some default values for fields we don't really have in melange. I'll think of maybe hm, just making one helper for building any `License` struct, but was worried as we'd need to make some function arguments optional.

Reply:
Actually, removed the function altogether! I think now it's actually better!